Report by day

ukrutt opened the pull request:
I need to report hours spent per day to a customer, and I thought I'd add a couple of switches and report machinery to accomplish this.

I based my report module on the existing `projects_section.py`.  There are a couple of possibly sticky/controversial points:

 - When using the new `--per-day` switch, the "Per Day" section will be printed *instead of* the Projects section.  I don't know if this fits with your vision of the tool; I'd be equally happy if the Per Day were printed *in addition to* Projects.

 - When reporting the duration, I report it as a decimal.  This makes it easier to paste into e.g. Excel of Google Sheets.  Moreover, I made the decision to **round up to nearest 1/10th of an hour**.  This seems to be a common practice, however I'm not certain if this approach should be hardcoded.  To avoid too much machinery I chose to hardcode it.

 - In addition to the `--per-day` switch I added a `--project` switch that will filter on a project name.  This operates quite independently of the `--per-day` machinery, however I need both.  It should be straightforward to split this into 2 PRs if you'd prefer (one could simply cherry-pick the very first commit on this branch to get the `--project` filter.)

I imagine that this would also need some tests; however my personal need for this is quite immediate so I chose to submit what I have for now.

larose commented:
Thank you @ukrutt, I will review your PRs soon!

ukrutt commented:
Thank you @larose !  As I have mentioned in some of the PRs, some of them might "conflict" with each other.  If you have any trouble consolidating, I'll be happy to do so!  Also, it might be that some of them conflict with your vision/overall architecture.  For one, I am not familiar with the IoC approach that you are using.

ukrutt commented:
@larose - it would be great if you expand the development instructions a little bit.  None of the `make` commands work for me due to missing packages -- my hunch is I have to do  something like `pipenv install --dev` first -- but `pipenv` is very confusing to me so I'm not sure.  (I've used it on a project before but stopped using it when it insisted on updating irrelevant stuff even when I used `pipenv install --keep-outdated <new-package>`...)  (FWIW, I did `pip install -e .` to get started.)

ukrutt commented on utt/report/model.py:
```diff
@@ -1,18 +1,22 @@
 
 class Report:
-    def __init__(self, activities, start_date, end_date, local_timezone):
+    def __init__(self, activities, start_date, end_date, local_timezone, args):
```
I could!

However: In my other PR #24 ("CSV output", which is a descendant of this PR #21) I also pass the `csv_section` parameter via `args`.  And in PR #23 ("details flag", which fixes #17) I pass the parameter `details` via `args`.  In a case like this, where there's an increasing number of parameters that we'd want to pass (all from command line switches), it seems cleaner to keep them collected.

(But I will do it the way you prefer.)

ukrutt commented on utt/commands/report.py:
```diff
@@ -50,5 +50,17 @@ def add_args(parser):
+            "--per-day",
+            action='store_true',
+            default=False,
+            help="Show total hours per per day.")
```
Fixed in 06f5c7acb269126ad67f48ef36713c7231359e4c

ukrutt commented:
@larose in 5ccfe45 and in 101f8c1 I have added integration tests for the `--per-day` and `--project <foo>` switches (and the combination of the two, which is I envision this feature.)

larose commented:
Thanks a lot for the tests @ukrutt! I'll merge your PR, but there is one last thing I'd like to check with you. Should we add `h` next to the billing hours? Actual output:

```

------ Monday, Aug 20, 2018 (week 34) to Tuesday, Aug 21, 2018 (week 34) -------

Working Time: 1h00
Break   Time: 0h00

----------------------------------- Per Day ------------------------------------

2018-08-20:  0.4  (0h24) - project_1, project_2 - task_1, task_2
2018-08-21:  0.6  (0h36) - project_1, project_2 - task_1, task_2, task_3

---------------------------------- Activities ----------------------------------

(0h12) project_1: task_1
(0h12) project_1: task_2
(0h06) project_1: task_3
(0h12) project_2: task_1
(0h12) project_2: task_2
(0h06) project_2: task_3

```

Proposed output:

```

------ Monday, Aug 20, 2018 (week 34) to Tuesday, Aug 21, 2018 (week 34) -------

Working Time: 1h00
Break   Time: 0h00

----------------------------------- Per Day ------------------------------------

2018-08-20:  0.4h  (0h24) - project_1, project_2 - task_1, task_2
2018-08-21:  0.6h  (0h36) - project_1, project_2 - task_1, task_2, task_3

---------------------------------- Activities ----------------------------------

(0h12) project_1: task_1
(0h12) project_1: task_2
(0h06) project_1: task_3
(0h12) project_2: task_1
(0h12) project_2: task_2
(0h06) project_2: task_3

```

Note the `h` next to `0.4` and `0.6`.

larose commented:
Also, if you can squash all you commits into one with a good commit message that would be awesome (I can do it if you want).

ukrutt commented:
I can definitely squash the commits.  My habit is to make *many* commits but I can see how it ends up becoming unwieldy.

Re: the `h` suffix, I agree that makes sense, and it's also easy.  One question: In #24 (CSV output) I'd prefer *not* to output the `h` suffix.  The plain number is easier to import into a spreadsheet.  Do you agree?  (I suppose that discussion is more appropriate to leave for the other PR.)

ukrutt commented:
Another note: I plan to change the rounding so that it rounds to *nearest* rather than rounding up -- except for the first 6 minutes, where we *always* round up.

Thus,

 - 0-8 min -> 0.1 h (NOTE: rounding up to 0.1 h for the first 5 minutes)
 - 9-14 min -> 0.2 h
 - 15-20 min -> 0.3 h
 :
 - 51-56 min -> 0.9 h
 - 57-62 min -> 1.0 hours
 - 63-68 min -> 1.1 hours
 - etc.

Any concerns or objections about this?  (It would be nice to make this configurable, but that's beyond the scope, and again, I'm catering to my personal need/preference.)

larose commented:
> Re: the `h` suffix, I agree that makes sense, and it's also easy. One question: In #24 (CSV output) I'd prefer _not_ to output the `h` suffix. The plain number is easier to import into a spreadsheet. Do you agree? (I suppose that discussion is more appropriate to leave for the other PR.)

Yes, I agree.

larose commented:
> Another note: I plan to change the rounding so that it rounds to _nearest_ rather than rounding up -- except for the first 6 minutes, where we _always_ round up.
> 
> Thus,
> 
> * 0-8 min -> 0.1 h (NOTE: rounding up to 0.1 h for the first 5 minutes)
> * 9-14 min -> 0.2 h
> * 15-20 min -> 0.3 h
>   :
> * 51-56 min -> 0.9 h
> * 57-62 min -> 1.0 hours
> * 63-68 min -> 1.1 hours
> * etc.
> 
> Any concerns or objections about this? (It would be nice to make this configurable, but that's beyond the scope, and again, I'm catering to my personal need/preference.)

That's fine with me!

ukrutt commented:
I've added 'h' to the billing hours as discussed.  I also updated the rounding.  I decided to round down for 1 and 2 minutes because it would overly complicated and seemed nitpicky.  I had to add a small hack (add 1e-6) to avoid "banker's rounding" -- the alternative would be to use the Decimal module as suggested here: https://stackoverflow.com/a/33019948/3061818.

If you're happy with this and want to merge, could you do the "squash to one commit"?  The commit message could be something like,

> Add switches to report for specific project, and also hours worked per day.

larose commented:
This PR has been squashed and merged manually. Thank you!

larose commented:
This has been released in v1.13.

larose commented:
> @larose - it would be great if you expand the development instructions a little bit. None of the `make` commands work for me due to missing packages -- my hunch is I have to do something like `pipenv install --dev` first -- but `pipenv` is very confusing to me so I'm not sure. (I've used it on a project before but stopped using it when it insisted on updating irrelevant stuff even when I used `pipenv install --keep-outdated <new-package>`...) (FWIW, I did `pip install -e .` to get started.)
> 
> UPDATE: I put my `.venv` aside and started fresh with `pipenv install --dev`. Was able to run the tests and have now passed both `make format`, `make lint`, and `make test`.
> 
> Will write some integration tests... does this mean adding to `utt/test/integration/Makefile`?

@ukrutt, I just wanted to let you know that I've improved the dev instructions: https://github.com/larose/utt/blob/master/docs/HACKING.md . It should be bit less confusing  :)

ukrutt commented:
Yes I saw -- thanks!
